solidity/contracts/ValidatorRegistration.sol

```diff
@@ -1,23 +1,113 @@
+// Adapted from: https://github.com/prysmaticlabs/prysm/blob/master/contracts/validator-registration-contract/validator_registration.sol
 pragma solidity ^0.4.23;
+pragma experimental ABIEncoderV2;
 
 contract ValidatorRegistration {
 
-  event Deposit(bytes32 _pubkey, uint256 _withdrawal_shard_id, address _withdrawal_address, bytes32 _randao_commitment);
+    event Eth1Deposit(
+        bytes previousReceiptRoot,
+        bytes data,
+        uint totalDepositcount
+    );
 
-  mapping(bytes32 => bool) public used_pubkeys;
+    event ChainStart(
+        bytes receiptRoot,
+        bytes time
+    );
 
-  function deposit(
-    bytes32 _pubkey,
-    uint256 _withdrawal_shard_id,
-    address _withdrawal_address,
-    bytes32 _randao_commitment
-    ) public payable {
-      require(msg.value == 32 ether);
-      require(!used_pubkeys[_pubkey]);
+    uint public constant DEPOSIT_SIZE = 32 ether;
+    uint public constant DEPOSITS_FOR_CHAIN_START = 8;
+    uint public constant MIN_TOPUP_SIZE = 1 ether;
+    uint public constant GWEI_PER_ETH = 10 ** 9;
+    // Setting MERKLE_TREE_DEPTH to 16 instead of 32 due to gas limit
+    uint public constant MERKLE_TREE_DEPTH = 16;
+    uint public constant SECONDS_PER_DAY = 86400;
 
-      used_pubkeys[_pubkey] = true;
+    mapping (uint => bytes) public receiptTree;
+    uint public totalDepositCount;
 
-      emit Deposit(_pubkey, _withdrawal_shard_id, _withdrawal_address, _randao_commitment);
-  }
+    // When users wish to become a validator by moving ETH from
+    // 1.0 chain to the 2.0 chain, they should call this function
+    // sending along DEPOSIT_SIZE ETH and providing depositParams
+    // as a simple serialize'd DepositParams object of the following
+    // form: 
+    // {
+    //    'pubkey': 'int256',
+    //    'proof_of_possession': ['int256'],
+    //    'withdrawal_credentials`: 'hash32',
+    //    'randao_commitment`: 'hash32'
+    // }
+    function deposit(
+        bytes depositParams
+    )
+        public
+        payable
+    {
+        uint index = totalDepositCount + 2 ** MERKLE_TREE_DEPTH;
+        bytes memory msgGweiInBytes = toBytes(msg.value);
+        bytes memory timeStampInBytes = toBytes(block.timestamp);
+        bytes memory depositData = mergeBytes(mergeBytes(msgGweiInBytes, timeStampInBytes), depositParams);
 
-}
+        emit Eth1Deposit(receiptTree[1], depositParams, totalDepositCount);
+
+        receiptTree[index] = abi.encodePacked(keccak256(depositData));
```

Comment:
I don't think we need abi.encodePacked here, we are only hashing 1 item. please confirm if this is correct

Reply:
@noot This is should be correct. [It appears](https://solidity.readthedocs.io/en/v0.5.0/abi-spec.html#non-standard-packed-mode) that, since the keccak chunk is 32 bytes, no difference will result after packing. I'll look into this and make changes if no problems arise.

Reply:
Resolved this.